config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java

```diff
@@ -0,0 +1,95 @@
+package com.typesafe.config.impl;
+
+import com.typesafe.config.ConfigNode;
+
+import java.util.*;
+
+final class ConfigNodeComplexValue implements ConfigNode, ConfigNodeValue {
+    private ArrayList<ConfigNode> children;
+    final private LinkedHashMap<Path, Integer> map = new LinkedHashMap<>();
+    final ArrayList<Integer> keyValueIndexes;
+
+    ConfigNodeComplexValue(Collection<ConfigNode> children) {
+        this.children = new ArrayList(children);
+        keyValueIndexes = new ArrayList<Integer>();
+
+        // Construct the list of indexes of Key-Value nodes. Do this
+        // in reverse order, since all but the final duplicate will be removed.
+        for (int i = this.children.size() - 1; i >= 0; i--) {
+            ConfigNode currNode = this.children.get(i);
+            if (currNode instanceof ConfigNodeKeyValue) {
+                keyValueIndexes.add(i);
+            }
+        }
+    }
+
+    public ArrayList<ConfigNode> children() {
+        return children;
+    }
+
+    public String render() {
+        StringBuilder renderedText = new StringBuilder();
+        for (ConfigNode child : children) {
+            renderedText.append(child.render());
+        }
+        return renderedText.toString();
```

Comment:
I think this is essentially the same issue as the `ConfigValue` hierarchy; there we have `AbstractConfigValue` which is extended by all the concrete value types, and we usually use `AbstractConfigValue` internally because it can have impl-private methods on it. We also have `AbstractConfigObject extends AbstractConfigValue`.

There's a good chance I bet that your `AbstractConfigNodeValue` would end up having a method on it eventually, but if it doesn't it isn't the end of the world.

If you have some casts to start with we can always clean it up later, I wouldn't get too stuck on it for the moment. Things may become clearer as you get more of the PR done.